### PR TITLE
test: Add invariant tests for SumTree structure in PDPVerifier

### PR DESCRIPTION
**Pull request by Hany-Almnaem**
## Summary
This PR introduces a suite of tests to validate the correctness and stability of the `sumTreeCounts` structure in the `PDPVerifier` smart contract.

The tests ensure that each node in the SumTree maintains the invariant:

* `sumTreeCounts[setId][index] == sum of rootLeafCounts[setId][j]` for the correct range derived from the node’s height.

This directly addresses and resolves issue: [#45](https://github.com/FilOzone/pdp/issues/45)

### *Core Tests (Basic Coverage)*

* `testSumTreeInvariantAfterAddsAndRemovals()`  
  - Adds a controlled set of roots, removes a few, and asserts the sumtree structure.
* `assertSumTreeInvariant(uint256 setId)`  
  - Shared invariant validation used across tests.
* `testAdvancedSumTreeInvariant()`  
  - Simulates multiple add/remove rounds using dynamic values.
* `testFuzzedSumTreeInvariant(uint256 seed)`  
  - Uses seeded randomness to test with pseudo-random leaf counts and root IDs.

### *Advanced Suite (Tagged `[ADVANCED]`)*

Implemented in a separate `SumTreeEnhancedTest` contract:

* `testSumTreeStressTest()`  
  - Applies add/remove operations over several rounds with increasing load.
* `testSumTreeEdgeCases()`  
  - Verifies correctness under very small, very large, and alternating root sizes.
* `testSumTreePerformance()`  
  - Executes 100 add/remove operations and logs gas usage.
* `testSumTreeRandomized(uint256 seed)`  
  - Interleaves randomized add/remove logic with proving period progression.
* Shared internal `assertSumTreeInvariant()` with more detailed error messaging for diagnostics.

## Benefits

* Validates the structural integrity of the SumTree under normal and edge-case usage.
* Improves test coverage for internal tree logic, including depth-aware summation.
* Reduces regression risk from future updates to root addition/removal logic.
* Keeps invariant assertions reusable and centralized for clarity and reliability.

## Notes

* Tests pass locally using `forge test`
* No contract logic has been modified—only test coverage has been added.
* Code has been linted and follows existing style conventions.
* Advanced tests are separated to keep the default suite performant and focused.

Please let me know if you prefer these broken into separate PRs or test files.

### CODE REVIEW

File: test/PDPVerifier.t.sol [hunk 1]
```diff
@@ -2025,6 +2146,207 @@ contract PDPVerifierMigrateTest is Test {
...
+        }
+    }
+
+    function createFreshDataSet() internal returns (uint256) {
+        return pdpVerifier.addPieces{value: PDPFees.sybilFee()}(
+            NEW_DATA_SET_SENTINEL, address(listener), new Cids.Cid[](0), abi.encode(empty, empty)
+        );
+    }
+
+    function assertSumTreeInvariant(uint256 setId) internal view {
```
Comment (ZenGround0) [inline test/PDPVerifier.t.sol]: Let's not duplicate this entire function from [here](https://github.com/FilOzone/pdp/pull/166/files#diff-1fb8671b384ff94f32f76f2e1c3c90182c0655bcfd6a5c9a6c2663c11e307d22R1429).  Please refactor this somehow, probably with a shared inherited contract, and remove duplication.

File: test/PDPVerifier.t.sol [hunk 2]
```diff
@@ -1426,6 +1426,127 @@ contract SumTreeAddTest is MockFVMTest, PieceHelper {
         assertEq(pdpVerifier.getDataSetLeafCount(testSetId), 820, "Incorrect final data set leaf count");
     }
 
+    function assertSumTreeInvariant(uint256 setId) internal view {
+        uint256 nextPieceId = pdpVerifier.getNextPieceId(setId);
+        for (uint256 index = 0; index < nextPieceId; index++) {
+            uint256 height = pdpVerifier.getTestHeightFromIndex(index);
+            uint256 range = 1 << height;
+
+            if (index + 1 < range) {
```
Comment (ZenGround0) [inline test/PDPVerifier.t.sol]: We should definitely not be swallowing these errors!  If something is this wrong with either the sumtree or the test then we need to fail and fix what is broken.

File: test/PDPVerifier.t.sol [hunk 3]
```diff
@@ -1426,6 +1426,127 @@ contract SumTreeAddTest is MockFVMTest, PieceHelper {
...
+            );
+        }
+
+        uint256 nextPieceId = pdpVerifier.getNextPieceId(testSetId);
+        assertGt(nextPieceId, numPieces, "New pieces not added properly");
+        assertGt(pdpVerifier.getDataSetLeafCount(testSetId), 0, "Leaf count should be positive after add/remove");
+    }
+
+    function testFuzzedSumTreeInvariant(uint256 seed) public {
+        vm.assume(seed > 100);
```
Comment (ZenGround0) [inline test/PDPVerifier.t.sol]: what's the thinking behind this?

File: test/PDPVerifier.t.sol [hunk 4]
```diff
@@ -2025,6 +2146,207 @@ contract PDPVerifierMigrateTest is Test {
...
+            // Process changes
+            pdpVerifier.nextProvingPeriod(testSetId, block.number + CHALLENGE_FINALITY_DELAY, empty);
+
+            // Verify invariants
+            assertSumTreeInvariant(testSetId);
+        }
+    }
+
+    /// @notice [ADVANCED] Tests SumTree behavior with edge case sizes and patterns
+    function testSumTreeEdgeCases() public {
```
Comment (ZenGround0) [inline test/PDPVerifier.t.sol]: I don't understand why this is "edge cases".  Again this seems like something we could remove without losing any conceptual coverage at all.

**Comment by BigLep** [comment]
@Hany-Almnaem : thank you for your work here, and apologies for the delay for having maintainers to engage.  Do you think you'll be able to incorporate comments so we can get this merged soon while the context is fresh?

**Review by ZenGround0 — APPROVED** [review]
Thx LGTM
